File: src/map_filter.cpp
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <rclcpp/rclcpp.hpp>

#include <memory>
#include <string>

namespace
{
class MapFilter : public rclcpp::Node
{
  public:
    MapFilter() : Node("map_filter")
    {
        declare_parameter<std::string>("input_topic", "/map");
        declare_parameter<std::string>("output_topic", "/map_valid");

        auto qos = rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable();
        const auto input_topic = get_parameter("input_topic").as_string();
        const auto output_topic = get_parameter("output_topic").as_string();

        publisher_ = create_publisher<nav_msgs::msg::OccupancyGrid>(output_topic, qos);
        subscription_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
            input_topic,
            qos,
            [this](nav_msgs::msg::OccupancyGrid::SharedPtr msg) {
                handle_map(*msg);
            });
    }

  private:
    void handle_map(const nav_msgs::msg::OccupancyGrid &msg)
    {
        if (msg.info.width == 0 || msg.info.height == 0)
        {
            ++dropped_empty_;
            if (dropped_empty_ == 1)
            {
                RCLCPP_WARN(get_logger(), "Dropping empty /map from the mapper");
            }
            return;
        }

        publisher_->publish(msg);
        if (!published_first_)
        {
            RCLCPP_INFO(
                get_logger(),
                "Publishing filtered map on /map_valid (%ux%u, resolution %.3f)",
                msg.info.width,
                msg.info.height,
                msg.info.resolution);
            published_first_ = true;
        }
    }

    rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr publisher_;
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr subscription_;
    int dropped_empty_{0};
    bool published_first_{false};
};
} // namespace

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MapFilter>());
    rclcpp::shutdown();
    return 0;
}

File: ros_test/map_filter.py
#!/usr/bin/env python3
"""Republish only valid OccupancyGrid messages from the mapper."""

from __future__ import annotations

import rclpy
from nav_msgs.msg import OccupancyGrid
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy


def transient_map_qos() -> QoSProfile:
    """Use latched map QoS so late subscribers immediately receive the map."""
    return QoSProfile(
        depth=1,
        durability=DurabilityPolicy.TRANSIENT_LOCAL,
        reliability=ReliabilityPolicy.RELIABLE,
    )


class MapFilter(Node):
    """Drop empty maps that can confuse downstream navigation tools."""

    def __init__(self) -> None:
        """Create the input subscription and filtered map publisher."""
        super().__init__("map_filter")

        # Topics are parameters so the node can be reused with other mappers.
        self.declare_parameter("input_topic", "/map")
        self.declare_parameter("output_topic", "/map_valid")

        # These flags keep startup logs useful without repeating every message.
        self._dropped_empty = 0
        self._published_first = False

        input_topic = self.get_parameter("input_topic").value
        output_topic = self.get_parameter("output_topic").value
        qos = transient_map_qos()

        self._publisher = self.create_publisher(OccupancyGrid, output_topic, qos)
        self._subscription = self.create_subscription(
            OccupancyGrid,
            input_topic,
            self._handle_map,
            qos,
        )

    def _handle_map(self, msg: OccupancyGrid) -> None:
        """Forward non-empty occupancy grids and suppress empty placeholders."""
        if msg.info.width == 0 or msg.info.height == 0:
            self._dropped_empty += 1
            if self._dropped_empty == 1:
                self.get_logger().warning("Dropping empty /map from the mapper")
            return

        self._publisher.publish(msg)
        if not self._published_first:
            self.get_logger().info(
                "Publishing filtered map on /map_valid "
                f"({msg.info.width}x{msg.info.height}, resolution {msg.info.resolution:.3f})"
            )
            self._published_first = True


def main(args: list[str] | None = None) -> None:
    """Run the map filter node."""
    rclpy.init(args=args)
    node = MapFilter()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
